lib/src/main/kotlin/foundation/algorand/deterministicP256/DeterministicP256.kt

```diff
@@ -0,0 +1,145 @@
+ * the License.
+ */
+package foundation.algorand.deterministicP256
+
+import android.security.keystore.KeyProperties
+import cash.z.ecc.android.bip39.Mnemonics.MnemonicCode
+import java.nio.ByteBuffer
+import java.security.KeyPair
+import java.security.KeyPairGenerator
+import java.security.MessageDigest
+import java.security.SecureRandom
+import java.security.Signature
+import java.security.interfaces.ECPrivateKey
+import java.security.spec.*
+import javax.crypto.SecretKeyFactory
+import javax.crypto.spec.PBEKeySpec
+
+/**
+ * DeterministicP256 - a class that generates deterministic P-256 keypairs from a BIP39 phrase and a
+ * domain-specific origin and userId.
+ *
+ * For generating passkeys intended for FIDO2-based authentication to web services, in a
+ * deterministic manner that allows a user to regenerate the same keypair on different devices.
+ *
+ * 1) Start by generating a root seed from a BIP39 phrase using PBKDF2-HMAC-SHA512 with 600k
+ * iterations. This should only be run once per device, and the root seed should be stored securely.
+ * The mnemonic phrase should only be inputed once and then be discarded by the device.
+ *
+ * 2) Generate a domain-specific keypair from the root seed, origin, and userId. The origin is the
+ * domain of the service, and the userId is the user's unique identifier on that service. A counter
+ * can also be set in case it is pertinent to generate multiple passkeys for a service.
+ *
+ * 3) Sign a payload with the domain-specific keypair. The keypairs can be stored and retreived from
+ * storage using some secure storage mechanism.
+ */
+class DeterministicP256 {
+        /** genRootSeedWithBIP39 - wrapper around genRootSeed that validates the BIP39 phrase. */
+        fun genRootSeedWithBIP39(
+                        phrase: String,
+                        salt: ByteArray = "liquid".toByteArray(),
+                        iterationCount: Int = 6000000,
+                        keyLength: Int = 512
+        ): ByteArray {
+                MnemonicCode(phrase).validate()
+                return genRootSeed(phrase.toCharArray(), salt, iterationCount, keyLength)
+        }
+
+        /** genRootSeed - generates a root seed from a char array using PBKDF2-HMAC-SHA512. */
+        private fun genRootSeed(
+                        entropy: CharArray,
+                        salt: ByteArray,
+                        iterationCount: Int,
+                        keyLength: Int
+        ): ByteArray {
+                val keySpec: KeySpec = PBEKeySpec(entropy, salt, iterationCount, keyLength)
+                val keyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512")
+                return keyFactory.generateSecret(keySpec).encoded
+        }
+
+        /**
+         * genDomainSpecificKeypair - generates a domain-specific keypair from a root seed, origin,
+         * userid and counter
+         */
+        fun genDomainSpecificKeypair(
+                        rootSeed: ByteArray,
```

Comment:
Are we using directly the root seed? Couldn't we just pass it through a hashing iteration to not load into memory BIP39 seed at run-time?

Reply:
I call it "root seed" here in the sense that it is the root of the domain-specific pass keys.

The root seed is itself the result of the mnemonic passing through 600000 SHA512 HMAC iterations

Reply:
It all makes sense. But yeah, let's just change the name to avoid confusion